src/src/com/microsoft/aad/adal/TokenCacheAccessor.java

```diff
@@ -0,0 +1,271 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.aad.adal;
+
+import java.io.UnsupportedEncodingException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Internal class handling the interaction with {@link AcquireTokenSilentHandler} and {@link ITokenCacheStore}. 
+ */
+class TokenCacheAccessor {
+    private static final String TAG = TokenCacheAccessor.class.getSimpleName();
+    
+    private final ITokenCacheStore mTokenCacheStore;
+    private final String mAuthority;
+    
+    TokenCacheAccessor(final ITokenCacheStore tokenCacheStore, final String authority) {
+        if (tokenCacheStore == null) {
+            throw new IllegalArgumentException("tokenCacheStore");
+        }
+        
+        if (StringExtensions.IsNullOrBlank(authority)) {
+            throw new IllegalArgumentException("authority");
+        }
+        
+        mTokenCacheStore = tokenCacheStore;
+        mAuthority = authority;
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for regular token cache entry.  
+     */
+    TokenCacheItem getRegularTokenCacheItemWithAT(final String resource, final String clientId, final String user) {
+        final String cacheKey = CacheKey.createCacheKeyForRTEntry(mAuthority, resource, clientId, user);
+        return mTokenCacheStore.getItem(cacheKey);
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for MRRT token cache entry.  
+     */
+    TokenCacheItem getMRRTItem(final String clientId, final String user) {
+        final String cacheKey = CacheKey.createCacheKeyForMRRT(mAuthority, clientId, user);
+        return mTokenCacheStore.getItem(cacheKey);
+    }
+    
+    /**
+     * @return {@link TokenCacheItem} for FRT token cache entry.  
+     */
+    TokenCacheItem getFRTItem(final String familyClientId, final String user) {
+        if (StringExtensions.IsNullOrBlank(user)) {
+            return null;
+        }
+        
+        final String cacheKey = CacheKey.createCacheKeyForFRT(mAuthority, familyClientId, user);
+        return mTokenCacheStore.getItem(cacheKey);
+    }
+    
+    /**
+     * Remove existing token cache item if needed and update token cache with returned auth result. 
+     */
+    void updateCachedItemToResult(final String resource, final String clientId, final AuthenticationResult result, 
+            final TokenCacheItem cachedItem) {
+        if (result == null) {
+            return;
+        }
+        
+        // remove Item if oauth2_error is invalid_grant
+        if (AuthenticationConstants.OAuth2ErrorCode.INVALID_GRANT.equalsIgnoreCase(result.getErrorCode())) {
```

Comment:
if errorcase...
 if it's invalid_grant
      delete
else successcase
    update